src/magic_folder/_schema.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+This module defines the database schema used by the model interface.
+
+:var int MAXIMUM_UPGRADES: The maximum number of upgrades that are allowed in
+    a single schema.
+"""
+
+from __future__ import (
+    unicode_literals,
+    absolute_import,
+)
+
+import attr
+
+_CREATE_VERSION = (
+    """
+    CREATE TABLE IF NOT EXISTS [schema-version] AS SELECT 0 AS [version]
+    """
+)
+
+_INCREMENT_VERSION = (
+    """
+    UPDATE [schema-version]
+    SET [version] = [version] + 1
+    """
+)
+
+_READ_VERSION = (
+    """
+    SELECT [version] FROM [schema-version]
+    """
+)
+
+MAXIMUM_UPGRADES = 1000
+
+@attr.s(frozen=True)
+class DatabaseSchemaTooNew(Exception):
+    """
+    The schema in the database is newer than the Python schema representation.
+
+    This version of the software cannot use this version of the database.
+    """
+    software_version = attr.ib()
+    database_version = attr.ib()
+
+
+@attr.s(frozen=True)
+class SchemaUpgrade(object):
+    """
+    An upgrade from one schema version to the next.
+
+    :ivar list[unicode] statements: A list of statements to execute to
+        complete this upgrade.
+    """
+    statements = attr.ib(validator=attr.validators.instance_of(list))
+
+    def run(self, cursor):
+        """
+        Execute this upgrade against the given cursor.
+
+        This method does no transaction management.  It uses the cursor in
+        whatever state it is in.
+
+        :param cursor: A DB-API cursor to use to run the SQL.
+        """
+        for statement in self.statements:
+            cursor.execute(statement)
+        cursor.execute(_INCREMENT_VERSION)
+
+@attr.s
+class Schema(object):
+    """
+    The schema for a single database.
+
+    :ivar list[SchemaUpgrade] upgrades: A list of schema upgrades.  Each
+        element upgrades the schema *from* the schema version with a number
+        corresponding to the index of that element.
+    """
+    upgrades = attr.ib()
+
+    @property
+    def version(self):
+        """
+        Get the version number which identifies this particular schema.
+        """
+        return len(self.upgrades)
+
+    @upgrades.validator
+    def _validate_upgrades(self, attribute, value):
+        if len(value) > MAXIMUM_UPGRADES:
+            raise ValueError(
+                "Schema has {} upgrades, greater than maximum allowed {}".format(
+                    len(value),
+                    MAXIMUM_UPGRADES,
```

Comment:
yea, pragmatic.  1000 is basically arbitrary.  seems like a huge number of schema upgrades, but not astronomical.  if we hit it, we'll implement some kind of truncation or compaction thing.  Either let `.upgrades[N]` be the upgrader for `M+N` instead of `N` (so you can delete your very-old schema upgraders) or let you take all your upgraders and smash them into `.upgrades[0]`.  Or maybe we just decide to raise the limit at that point.